### PR TITLE
New floating-to-decimal formatting routine

### LINKED ISSUE #7030: Printing of the largest and other very large floats is wrong
If you run std::float::infinity.next_after(1.0)` you get the following (unexpected result)

~~~
=> 179769313486231680088648464220646842686668242844028646442228680066046004606080400844208228060084840044686866242482868202680268820402884062800406622428864666882406066422426822086680426404402040202424880224808280820888844286620802664406086660842040886824002682662666864246642840408646468824200860804260804068888.0
~~~

the expected result is actually something like

~~~
=> 179769313486231580788648464220646842686668242844028646442228680066046004606080400844208228060084840044686866242482868202680268820402884062800406622428864666882406066422426822086680426404402040202424880224808280820888844286620802664406086660842040886824002682662666864246642840408646468824200860804260804068887.0
~~~

### PR DESCRIPTION
**Pull request by lifthrasiir**
This is a direct port of my prior work on the float formatting. The detailed description is available [here](https://github.com/lifthrasiir/rust-strconv#flt2dec). In brief,
- This adds a new hidden module `core::num::flt2dec` for testing from `libcoretest`. Why is it in `core::num` instead of `core::fmt`? Because I envision that the table used by `flt2dec` is directly applicable to `dec2flt` (cf. #24557) as well, which exceeds the realm of "formatting".
- This contains both Dragon4 algorithm (exact, complete but slow) and Grisu3 algorithm (exact, fast but incomplete).
- The code is accompanied with a large amount of self-tests and some exhaustive tests. In particular, `libcoretest` gets a new dependency on `librand`. For the external interface it relies on the existing test suite.
- It is known that, in the best case, the entire formatting code has about 30 KBs of binary overhead (judged from strconv experiments). Not too bad but there might be a potential room for improvements.

This is rather large code. I did my best to comment and annotate the code, but you have been warned.

For the maximal availability the original code was licensed in CC0, but I've also dual-licensed it in MIT/Apache as well so there should be no licensing concern.

This is [breaking-change] as it changes the float output slightly (and it also affects the casing of `inf` and `nan`). I hope this is not a big deal though :)

Fixes #7030, #18038 and #24556. Also related to #6220 and #20870.
## Known Issues
- [x] I've yet to finish `make check-stage1`. It does pass main test suites including `run-pass` but there might be some unknown edges on the doctests.
- [ ] Figure out how this PR affects rustc.
- [ ] Determine which internal routine is mapped to the formatting specifier. Depending on the decision, some internal routine can be safely removed (for instance, currently `to_shortest_str` is unused).

### CODE REVIEW

**Comment by pnkfelix** [comment]
r? pnkfelix

**Comment by lifthrasiir** [comment]
I've successfully finished `make check-stage1`. Also, the basic size benchmark done with `fn main() { println!("hello, pi is {}", std::f64::consts::PI) }`:

Options | 2015-04-19 Nightly | This PR | Delta
-------- | --------:| --------:| --------:
None | 583,741 | 524,126 | -59,615
`-O` | 583,501 | 523,887 | -59,614
`-O -C lto` | 536,769 | 474,524 | -62,245
`-C opt-level=3 -C lto -C link-args=-s` | 294,128 | 314,584 | +20,456

It seems that the original float formatting code had 10KB of overhead, so the best case overhead is somewhat decreased.

**Comment by pnkfelix** [comment]
Can you here summarize the effect of "it changes the float output slightly" beyond "affects the casing of inf and nan")?

In particular: There is some discussion in other tickets about things like:
* whether to not to include the minus sign on `-0.0`,
* or likewise, whether to include the `.0` in `+/-0.0`.

I suspect that *those* sorts of changes are in some ways *more* breaking than merely fixing our bugs with round-off error on the actual non-zero digits that we emit.  And so it would be useful to know up front here whether this PR changes any of those behaviors, or if the changes to the float output is restricted to which (and how many) digits it chooses to emit on non-zero values.

**Comment by lifthrasiir** [comment]
@pnkfelix It is a bit hard to say, precisely because the old behavior is not well-behaved. I can however surely say that this does not change the intention (say, it does't make `{:e}` to print decimals only). More precisely, I *think* that exact changes are as follows:

- When the precision is not given, all specifiers default to the shortest representation.
- When the precision is given, all specifiers rounds precisely based on the exact (not shortest) decimal expansion. Unit tests have some examples, but namely, the exact value of `0.95f64` is slightly smaller than 0.95 and `format!("{:.1}", 0.95f64)` will give `"0.9"` instead of `"1.0"`.
- Given the enough precision (more than 800 fractional digits) it will (correctly) print trailling zeroes.
- Every letter inside the formatted number (`e`, `inf`, `nan`) is lowercased except for `{:E}` where it is uppercased instead.
- `nan` is unsigned, i.e. `format!("{:+?}", 0.0/0.0)` is simply `"nan"`.

Regarding the sign, with an exception of `nan`, it behaves identically as before. It is easy to switch the behavior (e.g. in order to get a trailing `.0`, use `frac_digits` of 1), so I'm not too worried.

File: src/libcore/num/flt2dec/mod.rs [hunk 1]
```diff
@@ -0,0 +1,662 @@
...
+We will call this mode of operation as to the *shortest* mode. This mode is used
+when there is no additional constraint, and can be thought as a "natural" mode
+as it matches the ordinary intuition (it at least prints `0.1f32` as "0.1").
+
+We have two more modes of operation closely related to each other. In these modes
+we are given either the number of significant digits `n` or the last-digit
+limitation `limit` (which determines the actual `n`), and we would like to get
+the representation `V = 0.d[0..n-1] * 10^k` such that:
+
+- `d[0]` is non-zero, unless `n` was zero in which case only `k` is returned.
```
Comment (pnkfelix) [inline src/libcore/num/flt2dec/mod.rs]: I don't understand how the returned value for `k` is meaningful in the case where `n` is 0 ... is it implicit that in such a scenario, the returned `d[0]` *would* have been non-zero if `n` had been positive?

Comment (lifthrasiir) [inline src/libcore/num/flt2dec/mod.rs]: No. For example, `v = 3.14` and `limit = 3` gives `n = 0` and `k = 3`. The extrapolated digits would start with `00314...`.

Comment (lifthrasiir) [inline src/libcore/num/flt2dec/mod.rs]: Ah, re-reading your comment I understand what you really meant; yes, `k <= limit` holds if and only if `n = 0`. There actually was one wrong comment (derived from the older faulty implementation) on `to_exact_fixed_str`. I'll fix that soon.

File: src/libcore/num/flt2dec/mod.rs [hunk 2]
```diff
@@ -0,0 +1,662 @@
...
+
+- `format_shortest(decoded, buf)`, which always needs at least
+  `MAX_SIG_DIGITS` digits of buffer. Implements the shortest mode.
+
+- `format_exact(decoded, buf, limit)`, which accepts as small as
+  one digit of buffer. Implements exact and fixed modes.
+
+They try to fill the `u8` buffer with digits and returns the number of digits
+written and the exponent `k`. They are total for all finite `f32` and `f64`
+inputs (Grisu internally falls back to Dragon if possible).
```
Comment (pnkfelix) [inline src/libcore/num/flt2dec/mod.rs]: I assume you meant to write "falls back to Dragon if *necessary*", yes?

Comment (lifthrasiir) [inline src/libcore/num/flt2dec/mod.rs]: Ah, yes.

File: src/libcore/num/flt2dec/mod.rs [hunk 3]
```diff
@@ -0,0 +1,662 @@
...
+- `to_exact_exp_str` prints the exact representation with given number of
+  digits in the exponential form.
+
+- `to_exact_fixed_str` prints the fixed representation with *exactly*
+  given number of fractional digits.
+
+They all return a slice of preallocated `Part` array, which corresponds to
+the individual part of strings: a fixed string, a part of rendered digits,
+a number of zeroes or a small (`u16`) number. The caller is expected to
+provide an enough buffer and `Part` array, and to assemble the final
```
Comment (pnkfelix) [inline src/libcore/num/flt2dec/mod.rs]: I think you left out a word here, perhaps you intended "The caller is expected to provid *a large* enough buffer and `Part` array, ..."

Comment (lifthrasiir) [inline src/libcore/num/flt2dec/mod.rs]: You are right.

**Comment by rprichard** [comment]
The `Display` behavior prior to this change always used a decimal format (like `%f`). With this PR, it uses either exponent or decimal depending on magnitude (like `%g`). I think that's the right default, which #24556 argues for, but it's also a rather largish breaking change that I didn't see mentioned above. `Display` can format `1e300`, which breaks the JSON serializer:

[JSON](https://github.com/rust-lang/rustc-serialize/blob/031333c08d4819aa36127f1a4070e3efcc2341d7/src/json.rs#L519):
```rust
            let s = v.to_string();
            if s.contains(".") {s} else {s + ".0"}
```
`1e300.0` is not valid.

The CSV serializer works because `Display` in this PR reverts to decimal when a precision is specified:

[CSV](https://github.com/BurntSushi/rust-csv/blob/07fbf494f3d0c8a1cabf57ded2aa4b93dd171857/src/encoder.rs#L164):
```rust
    let s: String = format!("{:.10}", v).trim_right_matches('0').into();
    if s.ends_with('.') { s + "0" } else { s }
```

**Comment by lifthrasiir** [comment]
@rprichard Aargh, I forgot to mention that! I have updated the earlier comment to include that breaking change. Thank you!

(As mentioned above, this too is very easy to change though. I explicitly mentioned this kind of decision above to gather the future direction, as it seems that there are ongoing discussions about changing defaults.)

**Comment by pnkfelix** [comment]
@lifthrasiir hmm, it will probably be easier to land the most important parts of this PR (namely, the *bug fixes*) if you remove a breaking change like the one that @rprichard mentioned.

But, I'll make sure we talk about this at the team meeting tonight.  Maybe we can go ahead and change this without going through an RFC.

**Comment by lifthrasiir** [comment]
@pnkfelix Yeah, that was my oversight, I forgot to replicate the old behavior in some cases. I'll rebase this tonight (i.e. within 6 hours) to avoid any breaking changes except for the rounding and inaccurate result.

**Comment by rprichard** [comment]
FWIW, the `%g`-like format is the default in many languages -- probably because it's non-lossy, concise, and convenient for values near zero in magnitude.  It *is* awkward that it can take either of two different forms, though, and perhaps it's too late.  I think the precision field for `%g` typically controls the total number of digits (before and after the decimal point).

File: src/libcore/num/flt2dec/bignum.rs [hunk 4]
```diff
@@ -0,0 +1,357 @@
...
+
+                // shift by `digits * digitbits` bits
+                for i in (0..self.size).rev() {
+                    self.base[i+digits] = self.base[i];
+                }
+                for i in 0..digits {
+                    self.base[i] = 0;
+                }
+
+                // shift by `nbits` bits
```
Comment (pnkfelix) [inline src/libcore/num/flt2dec/bignum.rs]: nit: `nbits` must be referring to `bits` here, right?

Comment (lifthrasiir) [inline src/libcore/num/flt2dec/bignum.rs]: Yes.

File: src/libcore/num/flt2dec/bignum.rs [hunk 5]
```diff
@@ -0,0 +1,357 @@
...
+                            sz += 1;
+                        }
+                        if retsz < i + sz {
+                            retsz = i + sz;
+                        }
+                    }
+                    retsz
+                }
+
+                let mut ret = [0; $n];
```
Comment (pnkfelix) [inline src/libcore/num/flt2dec/bignum.rs]: out of curiosity, would it be potentially faster here to instead copy `self.base` into the stack-allocated temp, and write the output of `mul_inner` into `self.base` directly?

(it seems like either way one has a copy of an array of length `$n`, so maybe it makes no difference... just curious)

Comment (lifthrasiir) [inline src/libcore/num/flt2dec/bignum.rs]: Good observation. I would check if this would make a difference.

File: src/libcore/num/flt2dec/decoder.rs [hunk 6]
```diff
@@ -0,0 +1,98 @@
...
+            // so the mantissa is scaled for subnormals.
+            FullDecoded::Finite(Decoded { mant: mant, minus: 1, plus: 1,
+                                          exp: exp, inclusive: even })
+        }
+        FpCategory::Normal => {
+            let minnorm = <T as DecodableFloat>::min_pos_norm_value().integer_decode();
+            if mant == minnorm.0 && exp == minnorm.1 {
+                // (maxmant, exp - 1) -- (minnormmant, exp) -- (minnormmant + 1, exp)
+                // where maxmant = minnormmant * 2 - 1
+                FullDecoded::Finite(Decoded { mant: mant << 1, minus: 1, plus: 2,
```
Comment (pnkfelix) [inline src/libcore/num/flt2dec/decoder.rs]: I'm having trouble connecting up the above comment with this code here.

IIUC, your notation is lower_bound_for_rounding -- value -- upper_bound_for_rounding

in the above, we are in the branch where `mant == minnorm`, so I am thinking that the uses of `mant` in the code match up with the uses of `minnormmant` in the comment.

And then you define `maxmant` to a value equivalent to `(mant << 1) - 1`.

But then you use `maxnant` in only the lower_bound_for_rounding in the comment, while from the code, it seems like the `mant << 1` value is relevant to the other two parts too: value and upper_bound_for_rounding... I guess that's the place where I'm confused; why aren't the value and upper_bound_for_rounding defined to be `minnormant * 2` and `minnormant * 2 + 1`

File: src/libcore/num/flt2dec/estimator.rs [hunk 7]
```diff
@@ -0,0 +1,23 @@
...
+
+/// Finds `k_0` such that `10^(k_0-1) < mant * 2^exp <= 10^(k_0+1)`.
+///
+/// This is used to approximate `k = ceil(log_10 (mant * 2^exp))`;
+/// the true `k` is either `k_0` or `k_0+1`.
+#[doc(hidden)]
+pub fn estimate_scaling_factor(mant: u64, exp: i16) -> i16 {
+    // 2^(nbits-1) < mant <= 2^nbits if mant > 0
+    let nbits = 64 - (mant - 1).leading_zeros() as i64;
+    (((nbits + exp as i64) * 1292913986) >> 32) as i16
```
Comment (pnkfelix) [inline src/libcore/num/flt2dec/estimator.rs]: Where does this formula come from?  Especially the magic constant 1292913986?  I know there's lot of doc in other modules, but it probably would be good to put a reference here.

File: src/libcore/num/flt2dec/decoder.rs (hunk 6 above)
Comment (lifthrasiir) [inline src/libcore/num/flt2dec/decoder.rs]: Uh, this actually is a bug: the range was twice large. This should read `mant: mant << 2, minus: 1, plus: 2, exp: exp - 2` (so that, the bounds are `(mant - 1/4) * 2^exp` through `(mant + 1/2) * 2^exp`). I don't know how this escaped my attention, but I'll add a test case for this.

The comment is correct by the way; the notation reads like `(value - 1 ulp) -- value -- (value + 1 ulp)`.

File: src/libcore/num/flt2dec/estimator.rs (hunk 7 above)
Comment (lifthrasiir) [inline src/libcore/num/flt2dec/estimator.rs]: That is `floor(2^32 * log_10 2)`. Thus it may underestimate the true `k`, but the approximated `k_0` cannot exceed it. (I agree that this needs a comment.)

File: src/libcore/num/flt2dec/mod.rs (hunk 1 above)
Comment (pnkfelix) [inline src/libcore/num/flt2dec/mod.rs]: No, I was and still am confused. In the comment above, `n` is the number of digits that we need to extract from the array `d` in the representation `V = 0.d[0..n-1] * 10^k`.  My comments were based on that equation; it led me to ask the question: *when* can `n` be zero?

At the time when I wrote that question, I had not yet read the paragraph that starts with
> When `limit` is given but not `n`, ...

which seems to describe scenarios that yield non-positive values for `n`.  But this mostly just confused me further, probably because I do not yet actually understand the fixed mode conversion.

The text in that paragraph that makes it sound like the equation `V = 0.d[0..n-1] * 10^k` simply cannot be well-defined ... maybe that is the source of my confusion.

I will try reading the whole text here and see if I can suggest a way to rephrase it to be clearer.

Comment (pnkfelix) [inline src/libcore/num/flt2dec/mod.rs]: (BTW I do recognize that the equation for `V` makes sense for `n = 0` when `V = 0` itself.  But you yourself gave the example of `v = 3.14`; so clearly I am still not understanding something here.)

File: src/libcore/num/flt2dec/decoder.rs (hunk 6 above)
Comment (lifthrasiir) [inline src/libcore/num/flt2dec/decoder.rs]: (The reason that this escaped my attention is that, the exhaustive check assumes that the decoder is correct, and there is no separate decoder test. Fortunately for us, there are only two values which can trigger this situation---`std::{f32, f64}::MIN_POSITIVE`---and they are explicitly tested.)

File: src/libcore/num/flt2dec/mod.rs (hunk 1 above)
Comment (lifthrasiir) [inline src/libcore/num/flt2dec/mod.rs]: Probably I've introduced the concepts backwards.

* The "exact mode conversion" is the mode that the caller demands the specified number of *significant* digits.
* The "fixed mode conversion" is the mode that the caller demands the specified number of *fractional* digits.

The fixed mode can be implemented via the exact mode and the estimator, which calculates the required number of significant digits from the number of fractional digits *and* the original value. It turns out that the internal digit generation code already has to estimate the exponent (which is almost same to the difference between them), so we can reuse this logic.

In order to achieve this merger, the combined exact-fixed mode is implemented so that both the number of significant digits `k` and the number of fractional digits `frac_digits` are given. (The actual interface differs from this to be more flexible: `k == buf.len()` and `frac_digits == -limit`. This doesn't change the discussion however.) The digit generation code will finish whichever comes first, so it is possible that there are no digits generated.

File: src/libcore/num/flt2dec/decoder.rs (hunk 6 above)
Comment (lifthrasiir) [inline src/libcore/num/flt2dec/decoder.rs]: Okay, I forgot to see that this affects many 2^n values. I've added a new commit which has both the fix and regression tests.

File: src/libcore/num/flt2dec/mod.rs (hunk 1 above)
Comment (pnkfelix) [inline src/libcore/num/flt2dec/mod.rs]: Oh wait!  Are you actually saying:

> we would like to get the representation `V = 0.d[0..n-1] * 10^k` such that `d[0]` is non-zero, unless `n` was zero in which case *only `k` digits are* returned.

*That* I could understand, I think!

Comment (lifthrasiir) [inline src/libcore/num/flt2dec/mod.rs]: Uh, no. `k` is just the exponent which does not depend on the determined `n` (which *is* the number of digits actually returned). Say that, for certain `v`, the algorithm returned `0.123 * 10^-8` (which implies that `n = 3`, and also implies either `len = 3` or `limit = -11`). Then for some other values for `limit`, it *may* return `0. * 10^-8` (with `n = 0` and `k = -8`). `10^-8` part still remains here, hence "we will only get `k`".

File: src/libcore/num/flt2dec/strategy/grisu.rs [hunk 8]
```diff
@@ -0,0 +1,749 @@
...
+use num::flt2dec::{Decoded, MAX_SIG_DIGITS, round_up};
+
+/// A custom 64-bit floating point type, representing `f * 2^e`.
+#[derive(Copy, Clone, Debug)]
+#[doc(hidden)]
+pub struct Fp {
+    /// The integer mantissa.
+    pub f: u64,
+    /// The exponent in base 2.
+    pub e: i16,
```
Comment (pnkfelix) [inline src/libcore/num/flt2dec/strategy/grisu.rs]: I haven't read the Grisu paper carefully, but in its discussion of the `diy_fp` struct, it notes that (1.) the paper's definition uses an unlimited range value for `e_x` to simplify the proofs, (2.) in practice the exponent type must only have "slightly greater range than the input exponent" and (3.) for IEEE doubles which have 11-bit exponents, a *32-bit* signed integer is by far big enough.

The combination of 2. and 3. leads me to wonder whether we could get into trouble using only `i16` here.  I know you exhaustively tested your algorithm for `f32`, but how confident are you that `i16` suffices for `f64` ?

Comment (lifthrasiir) [inline src/libcore/num/flt2dec/strategy/grisu.rs]: That is more than enough.

`diy_fp` is converted from the decoded float, then normalized, then multiplied by the cached power of 10 to get it between a certain interval (to be exact, `[4, 2^32)`). Since the normalization step can decrease the exponent by at most 63, and the original float has the minimal exponent of -1075, the minimum exponent possible is -1138.

File: src/libcoretest/num/flt2dec/bignum.rs [hunk 9]
```diff
@@ -0,0 +1,160 @@
...
+    assert_eq!(*Big::from_small(0).add(&Big::from_small(3)), Big::from_small(3));
+    assert_eq!(*Big::from_small(3).add(&Big::from_u64(0xfffe)), Big::from_u64(0x10001));
+    assert_eq!(*Big::from_u64(0xfedc).add(&Big::from_u64(0x789)), Big::from_u64(0x10665));
+    assert_eq!(*Big::from_u64(0x789).add(&Big::from_u64(0xfedc)), Big::from_u64(0x10665));
+}
+
+#[test]
+#[should_panic]
+fn test_add_overflow_1() {
+    Big::from_small(1).add(&Big::from_u64(0xffffff));
```
Comment (pnkfelix) [inline src/libcoretest/num/flt2dec/bignum.rs]: these tests are marked as `#[should_panic]`, but I cannot tell from reviewing the code where overflow would unconditionally originate.

Does the `panic` come from an index-out-of-bounds access, e.g. [here](https://github.com/rust-lang/rust/pull/24612/files#diff-f9782b6464182bbd9563eef49cc62bd7R130)?  Or does it come from an arithmetic-overflow, which means that one cannot rely on the `#[should_panic]` here actually occurring in general, not without `-C debug-assertions` turned on?

File: src/libcore/num/flt2dec/mod.rs [hunk 10]
```diff
@@ -0,0 +1,661 @@
...
+
+We will call the mode of operation requiring `n` as to the *exact* mode,
+and one requiring `limit` as to the *fixed* mode. The exact mode is a subset of
+the fixed mode: the sufficiently large last-digit limitation will eventually fill
+the supplied buffer and let the algorithm to return.
+
+# Implementation overview
+
+It is easy to get the floating point printing correct but slow (Russ Cox has
+[demonstrated](http://research.swtch.com/ftoa) how it's easy), or incorrect but
```
Comment (pnkfelix) [inline src/libcore/num/flt2dec/mod.rs]: Should we actually add an implementation of Russ Cox's code for purposes of comparative validation of the Dragon implementation, in the same way that you have validated your Grisu3 implementation against your Dragon4 implementation?

File: src/libcoretest/num/flt2dec/bignum.rs (hunk 9 above)
Comment (lifthrasiir) [inline src/libcoretest/num/flt2dec/bignum.rs]: Your guess is correct: for `add` and `mul_*` the `panic` should come from the out-of-bounds access. For `sub` it should come from `assert!(noborrow)`.

File: src/libcore/num/flt2dec/mod.rs (hunk 1 above)
Comment (pnkfelix) [inline src/libcore/num/flt2dec/mod.rs]: Oh sorry, somehow I didn't hit "reload" before I posted my last comment, so it may seem like I completely ignored you in my last comment.

I am re-reading the Steele+White Dragon paper now; I think your terminology matches theirs, so I just need to digest section 8 of their paper and figure out the right way to succinctly describe it.

Though to be honest the comment you wrote that points out the distinction between requesting a number of significant digits versus a number of fractional digits has helped me a lot.

Okay, back to reading, thanks.

File: src/libcore/num/flt2dec/mod.rs (hunk 10 above)
Comment (lifthrasiir) [inline src/libcore/num/flt2dec/mod.rs]: Russ Cox's code only deals with the exact mode (quite reasonably), so it is not directly comparable.

File: src/libcore/num/flt2dec/mod.rs (hunk 1 above)
Comment (lifthrasiir) [inline src/libcore/num/flt2dec/mod.rs]: You are welcome. Actually, I guess my comment above can directly go to the module documentation...

**Comment by pnkfelix** [comment]
(By the way this is all looking pretty fantastic so far.  I just want to really make sure I or someone else besides you has a decent understanding of all of it, so please do bear with me while I continue to work my way through it.)

**Comment by pnkfelix** [comment]
Also your note that "Depending on the decision, some internal routine can be safely removed (for instance, currently to_shortest_str is unused)." -- I would think that this may represent a deficiency in the expressiveness of our print formatting routines.  At least, it seemed like all of these variants were justified based on potential user needs in the Dragon paper...

File: src/libcore/num/flt2dec/bignum.rs (hunk 5 above)
Comment (lifthrasiir) [inline src/libcore/num/flt2dec/bignum.rs]: There was no significant difference (actually, that was very slightly slower).

**Comment by brson** [comment]
This is a really fun patch to read @lifthrasiir, and thanks for the thorough review @pnkfelix.

Here's the Grisu paper for those reading along http://www.cs.tufts.edu/~nr/cs257/archive/florian-loitsch/printf.pdf

**Comment by lifthrasiir** [comment]
Rebased.

**Comment by pnkfelix** [comment]
not accepted for beta.  (We all agreed that this is awesome, but also that it can be reasonably classified as a bug fix, and therefore can wait for the next release.)

File: src/libcore/fmt/mod.rs [hunk 11]
```diff
@@ -924,18 +1000,50 @@ impl<'a, T> Pointer for &'a mut T {
...
+}
+
+// Common code of floating point LowerExp and UpperExp.
+fn float_to_exponential_common<T>(fmt: &mut Formatter, num: &T, upper: bool) -> Result
+    where T: flt2dec::DecodableFloat
+{
+    let force_sign = fmt.flags & (1 << (FlagV1::SignPlus as u32)) != 0;
+    let sign = match force_sign {
+        false => flt2dec::Sign::Minus,
+        true  => flt2dec::Sign::MinusPlus,
```
Comment (sinistersnare) [inline src/libcore/fmt/mod.rs]: I don't see a lot of matching on booleans, should this be changed/does it matter?

Comment (pnkfelix) [inline src/libcore/fmt/mod.rs]: (i'm okay with it either way; i certainly wouldn't block this PR on changing this to an `if` expression)

Comment (sinistersnare) [inline src/libcore/fmt/mod.rs]: Sure, was curious nonetheless. Thanks for your input :)

Comment (lifthrasiir) [inline src/libcore/fmt/mod.rs]: @sinistersnare That is to be consistent with `float_to_decimal_common` code above. I have no strong feeling about this code, anyway.

File: src/libcore/num/flt2dec/mod.rs [hunk 12]
```diff
@@ -0,0 +1,661 @@
...
+/// it will only print given digits and nothing else.
+fn digits_to_dec_str<'a>(buf: &'a [u8], exp: i16, frac_digits: usize,
+                         parts: &'a mut [Part<'a>]) -> &'a [Part<'a>] {
+    assert!(!buf.is_empty());
+    assert!(buf[0] > b'0');
+    assert!(parts.len() >= 4);
+
+    // if there is the restriction on the last digit position, `buf` is assumed to be
+    // left-padded with the virtual zeroes. the number of virtual zeroes, `nzeroes`,
+    // equals to `max(0, exp + frag_digits - buf.len())`, so that the position of
```
Comment (pnkfelix) [inline src/libcore/num/flt2dec/mod.rs]: typo: "frac_digits"

Comment (lifthrasiir) [inline src/libcore/num/flt2dec/mod.rs]: Yup, I will fix that in the next rebase.

**Comment by lifthrasiir** [comment]
I've rebased the commit to account for another typo and banker's rounding in the exact mode. r? @pnkfelix

**Comment by pnkfelix** [comment]
@lifthrasiir okay, thanks for your patience with this.

I'm going to r+ this.  I cannot claim that I understand the algorithms 100%, but I have reviewed all the papers and satisfied myself that this implementation is a good match for them, and that the goals of the papers are a match for what we want in Rust.

**Comment by tshepang** [comment]
epic!

**Comment by lifthrasiir** [comment]
Whoops, I forgot to run `make check-stage1` (not just `make check-stage1-coretest`). This failure is nothing to do with the implementation but only the tests, I'll shortly fix that.

**Comment by lifthrasiir** [comment]
I've confirmed this new commit passes `make check-stage1`. r? @pnkfelix

**Comment by cmr** [comment]
Awesome! Epic review @pnkfelix, and thanks to @lifthrasiir for implementing these tricky algorithms.

**Comment by Kimundi** [comment]
Cool! Finally a proper implementation for this stuff :)
